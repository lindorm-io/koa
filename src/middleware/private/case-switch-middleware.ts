import { KoaContext, Middleware } from "../../typing";
import { camelKeys, isObjectStrict, snakeKeys } from "@lindorm-io/core";
import { merge } from "lodash";

export const caseSwitchMiddleware: Middleware<KoaContext> = async (ctx, next): Promise<void> => {
  ctx.params = isObjectStrict(ctx.params) ? camelKeys(ctx.params) : {};
  ctx.query = isObjectStrict(ctx.query) ? camelKeys(ctx.query) : {};
  ctx.request.body = isObjectStrict(ctx.request.body) ? camelKeys(ctx.request.body) : {};
  ctx.data = merge(ctx.params, ctx.query, ctx.request.body);

  await next();

  ctx.body = isObjectStrict(ctx.body) ? snakeKeys(ctx.body) : ctx.body;
};
