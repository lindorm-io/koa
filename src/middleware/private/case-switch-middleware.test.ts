import MockDate from "mockdate";
import { caseSwitchMiddleware } from "./case-switch-middleware";

MockDate.set("2020-01-01T08:00:00.000Z");

const next = () => Promise.resolve();

describe("bodyCaseSwitchMiddleware", () => {
  const array = ["array"];
  const date = new Date();
  const error = new Error("error");
  const string = "string";

  let ctx: any;

  beforeEach(() => {
    ctx = {
      request: {
        body: {
          camelCaseOne: "camelCaseOne",
          PascalCaseOne: "PascalCaseOne",
          snake_case_one: "snake_case_one",
          array,
          date,
          error,
          string,
        },
      },
      body: {
        camelCaseTwo: "camelCaseTwo",
        PascalCaseTwo: "PascalCaseTwo",
        snake_case_two: "snake_case_two",
        array,
        date,
        error,
        string,
      },
      params: {
        camelCaseThree: "camelCaseThree",
        PascalCaseThree: "PascalCaseThree",
        snake_case_three: "snake_case_three",
        array,
        date,
        error,
        string,
      },
      query: {
        camelCaseFour: "camelCaseFour",
        PascalCaseFour: "PascalCaseFour",
        snake_case_four: "snake_case_four",
        array,
        date,
        error,
        string,
      },
    };
  });

  test("should transform all incoming body to camelCase", async () => {
    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.request.body).toMatchSnapshot();
  });

  test("should transform all incoming params to camelCase", async () => {
    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.params).toMatchSnapshot();
  });

  test("should transform all incoming query to camelCase", async () => {
    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.query).toMatchSnapshot();
  });

  test("should merge all incoming data into one object on context", async () => {
    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.data).toMatchSnapshot();
  });

  test("should transform all outgoing body to snake_case", async () => {
    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.body).toMatchSnapshot();
  });

  test("should not fail if data is undefined", async () => {
    ctx = { params: undefined, request: { body: undefined }, body: undefined };

    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.params).toStrictEqual({});
    expect(ctx.request.body).toStrictEqual({});
    expect(ctx.body).toBeUndefined();
  });

  test("should not convert response body when not object", async () => {
    ctx = { params: "", query: "", request: { body: {} }, body: "string" };

    await expect(caseSwitchMiddleware(ctx, next)).resolves.toBeUndefined();

    expect(ctx.body).toBe("string");
  });
});
