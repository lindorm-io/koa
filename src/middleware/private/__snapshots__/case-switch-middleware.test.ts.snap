// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bodyCaseSwitchMiddleware should merge all incoming data into one object on context 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camelCaseFour": "camelCaseFour",
  "camelCaseOne": "camelCaseOne",
  "camelCaseThree": "camelCaseThree",
  "date": 2020-01-01T08:00:00.000Z,
  "error": [Error: error],
  "pascalCaseFour": "PascalCaseFour",
  "pascalCaseOne": "PascalCaseOne",
  "pascalCaseThree": "PascalCaseThree",
  "snakeCaseFour": "snake_case_four",
  "snakeCaseOne": "snake_case_one",
  "snakeCaseThree": "snake_case_three",
  "string": "string",
}
`;

exports[`bodyCaseSwitchMiddleware should transform all incoming body to camelCase 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camelCaseOne": "camelCaseOne",
  "date": 2020-01-01T08:00:00.000Z,
  "error": [Error: error],
  "pascalCaseOne": "PascalCaseOne",
  "snakeCaseOne": "snake_case_one",
  "string": "string",
}
`;

exports[`bodyCaseSwitchMiddleware should transform all incoming params to camelCase 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camelCaseFour": "camelCaseFour",
  "camelCaseOne": "camelCaseOne",
  "camelCaseThree": "camelCaseThree",
  "date": 2020-01-01T08:00:00.000Z,
  "error": [Error: error],
  "pascalCaseFour": "PascalCaseFour",
  "pascalCaseOne": "PascalCaseOne",
  "pascalCaseThree": "PascalCaseThree",
  "snakeCaseFour": "snake_case_four",
  "snakeCaseOne": "snake_case_one",
  "snakeCaseThree": "snake_case_three",
  "string": "string",
}
`;

exports[`bodyCaseSwitchMiddleware should transform all incoming query to camelCase 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camelCaseFour": "camelCaseFour",
  "date": 2020-01-01T08:00:00.000Z,
  "error": [Error: error],
  "pascalCaseFour": "PascalCaseFour",
  "snakeCaseFour": "snake_case_four",
  "string": "string",
}
`;

exports[`bodyCaseSwitchMiddleware should transform all outgoing body to snake_case 1`] = `
Object {
  "array": Array [
    "array",
  ],
  "camel_case_two": "camelCaseTwo",
  "date": 2020-01-01T08:00:00.000Z,
  "error": [Error: error],
  "pascal_case_two": "PascalCaseTwo",
  "snake_case_two": "snake_case_two",
  "string": "string",
}
`;
